File: takengine/mapengine/sdk/src/elevation/MultiplexingElevationChunkCursor.cpp
#include "elevation/MultiplexingElevationChunkCursor.h"

#include <algorithm>

#include "elevation/ElevationSource.h"

using namespace TAK::Engine::Elevation;

using namespace TAK::Engine::Feature;
using namespace TAK::Engine::Port;
using namespace TAK::Engine::Util;

namespace
{
    struct CursorComparator
    {
        CursorComparator(std::vector<bool(*)(ElevationChunkCursor &, ElevationChunkCursor &)> &impl_) :
            impl(impl_)
        {}

        bool operator()(const std::shared_ptr<ElevationChunkCursor> &a, const std::shared_ptr<ElevationChunkCursor> &b)
        {
            for (std::size_t i = 0u; i < impl.size(); i++) {
                const bool alt = impl[i](*a, *b);
                const bool blt = impl[i](*b, *a);
                if(alt) // strictly less
                    return true;
                else if(blt) // strictly greater
                    return false;
                // else, equal
            }
            return (intptr_t)a.get() < (intptr_t)b.get();
        }

        std::vector<bool(*)(ElevationChunkCursor &, ElevationChunkCursor &) NOTHROWS> &impl;
    };
}

MultiplexingElevationChunkCursor::MultiplexingElevationChunkCursor(Collection<std::shared_ptr<ElevationChunkCursor>> &cursors_) NOTHROWS
{
    do {
        TAKErr code(TE_Ok);

        this->order.push_back(ElevationSource_resolutionDesc);

        if (!cursors_.empty()) {
            Collection<std::shared_ptr<ElevationChunkCursor>>::IteratorPtr iter(nullptr, nullptr);
            code = cursors_.iterator(iter);
            TE_CHECKBREAK_CODE(code);

            do {
                std::shared_ptr<ElevationChunkCursor> element;
                code = iter->get(element);
                TE_CHECKBREAK_CODE(code);
                if (element->moveToNext() == TE_Ok)
                    this->cursors.push_back(element);
                code = iter->next();
                TE_CHECKBREAK_CODE(code);
            } while (true);
            if (code == TE_Done)
                code = TE_Ok;
            TE_CHECKBREAK_CODE(code);
        }

        if (!this->cursors.empty()) {
            CursorComparator comp(this->order);
            std::sort(this->cursors.begin(), this->cursors.end(), comp);
        }
    } while (false);
}

MultiplexingElevationChunkCursor::MultiplexingElevationChunkCursor(Collection<std::shared_ptr<ElevationChunkCursor>> &cursors_, bool(*order_)(ElevationChunkCursor &, ElevationChunkCursor &)) NOTHROWS
{
    do {
        TAKErr code(TE_Ok);
        if (!order_) {
            order_ = ElevationSource_resolutionDesc;
        }

        this->order.push_back(order_);


        if (!cursors_.empty()) {
            Collection<std::shared_ptr<ElevationChunkCursor>>::IteratorPtr iter(nullptr, nullptr);
            code = cursors_.iterator(iter);
            TE_CHECKBREAK_CODE(code);

            do {
                std::shared_ptr<ElevationChunkCursor> element;
                code = iter->get(element);
                TE_CHECKBREAK_CODE(code);
                if (element->moveToNext() == TE_Ok)
                    this->cursors.push_back(element);
                code = iter->next();
                TE_CHECKBREAK_CODE(code);
            } while (true);
            if (code == TE_Done)
                code = TE_Ok;
            TE_CHECKBREAK_CODE(code);
        }

        if (!this->cursors.empty()) {
            CursorComparator comp(this->order);
            std::sort(this->cursors.begin(), this->cursors.end(), comp);
        }
    } while (false);
}

MultiplexingElevationChunkCursor::MultiplexingElevationChunkCursor(Collection<std::shared_ptr<ElevationChunkCursor>> &cursors_, Collection<bool(*)(ElevationChunkCursor &, ElevationChunkCursor &)> &order_) NOTHROWS
{
    do {
        TAKErr code(TE_Ok);
        if (order_.empty()) {
            this->order.push_back(ElevationSource_resolutionDesc);
        } else {
            Collection<bool(*)(ElevationChunkCursor &, ElevationChunkCursor &)>::IteratorPtr iter(nullptr, nullptr);
            code = order_.iterator(iter);
            TE_CHECKBREAK_CODE(code);

            do {
                bool (*element)(ElevationChunkCursor &, ElevationChunkCursor &);
                code = iter->get(element);
                TE_CHECKBREAK_CODE(code);
                this->order.push_back(element);
                code = iter->next();
                TE_CHECKBREAK_CODE(code);
            } while (true);
            if (code == TE_Done)
                code = TE_Ok;
            TE_CHECKBREAK_CODE(code);
        }


        if (!cursors_.empty()) {
            Collection<std::shared_ptr<ElevationChunkCursor>>::IteratorPtr iter(nullptr, nullptr);
            code = cursors_.iterator(iter);
            TE_CHECKBREAK_CODE(code);

            do {
                std::shared_ptr<ElevationChunkCursor> element;
                code = iter->get(element);
                TE_CHECKBREAK_CODE(code);
                if(element->moveToNext() == TE_Ok)
                    this->cursors.push_back(element);
                code = iter->next();
                TE_CHECKBREAK_CODE(code);
            } while (true);
            if (code == TE_Done)
                code = TE_Ok;
            TE_CHECKBREAK_CODE(code);
        }

        if (!this->cursors.empty()) {
            CursorComparator comp(this->order);
            std::sort(this->cursors.begin(), this->cursors.end(), comp);
        }
    } while (false);
}

TAKErr MultiplexingElevationChunkCursor::get(ElevationChunkPtr &value) NOTHROWS
{
    return this->row->get(value);
}
TAKErr MultiplexingElevationChunkCursor::getResolution(double *value) NOTHROWS
{
    return this->row->getResolution(value);
}
TAKErr MultiplexingElevationChunkCursor::isAuthoritative(bool *value) NOTHROWS
{
    return this->row->isAuthoritative(value);
}
TAKErr MultiplexingElevationChunkCursor::getCE(double *value) NOTHROWS
{
    return this->row->getCE(value);
}
TAKErr MultiplexingElevationChunkCursor::getLE(double *value) NOTHROWS
{
    return this->row->getLE(value);
}
TAKErr MultiplexingElevationChunkCursor::getUri(const char **value) NOTHROWS
{
    return this->row->getUri(value);
}
TAKErr MultiplexingElevationChunkCursor::getType(const char **value) NOTHROWS
{
    return this->row->getType(value);
}
TAKErr MultiplexingElevationChunkCursor::getBounds(const Polygon2 **value) NOTHROWS
{
    return this->row->getBounds(value);
}
TAKErr MultiplexingElevationChunkCursor::getFlags(unsigned int *value) NOTHROWS
{
    return this->row->getFlags(value);
}
TAKErr MultiplexingElevationChunkCursor::moveToNext() NOTHROWS
{
    if(this->row.get()) {
        if(this->row->moveToNext() != TE_Ok)
            this->cursors.erase(this->cursors.begin());
        
        CursorComparator comp(this->order);
        std::sort(this->cursors.begin(), this->cursors.end(), comp);
    }
    if(this->cursors.empty())
        return TE_Done;
    this->row = this->cursors[0];
    return TE_Ok;
}

File: takkernel/engine/src/main/java/com/atakmap/map/elevation/MultiplexingElevationChunkCursor.java
package com.atakmap.map.elevation;

import com.atakmap.map.layer.feature.geometry.Geometry;
import com.atakmap.util.CascadingComparator;
import com.atakmap.util.Collections2;

import java.util.ArrayList;
import java.util.Collection;
import java.util.Collections;
import java.util.Comparator;
import java.util.List;

public final class MultiplexingElevationChunkCursor implements ElevationSource.Cursor {
    private ArrayList<ElevationSource.Cursor> cursors;
    private Comparator<ElevationSource.Cursor> order;
    private ElevationSource.Cursor row;

    public MultiplexingElevationChunkCursor(Collection<ElevationSource.Cursor> cursors) {
        this(cursors, Collections.<Comparator<ElevationSource.Cursor>>singletonList(ElevationSource.RESOLUTION_DESC));
    }

    public MultiplexingElevationChunkCursor(Collection<ElevationSource.Cursor> cursors, Comparator<ElevationSource.Cursor> order) {
        this(cursors, Collections.<Comparator<ElevationSource.Cursor>>singletonList(order));
    }

    public MultiplexingElevationChunkCursor(Collection<ElevationSource.Cursor> cursors, List<Comparator<ElevationSource.Cursor>> order) {
        Comparator<ElevationSource.Cursor> corder;
        if(order.size() == 1) {
            this.order = Collections2.first(order);
        } else if(order.isEmpty()) {
            this.order = ElevationSource.RESOLUTION_DESC;
        } else {
            this.order = new CascadingComparator<ElevationSource.Cursor>(order);
        }


        this.cursors = new ArrayList<ElevationSource.Cursor>(cursors.size());
        for(ElevationSource.Cursor c : cursors) {
            if(c.moveToNext())
                this.cursors.add(c);
        }
        Collections.sort(this.cursors, this.order);
    }

    @Override
    public ElevationChunk get() {
        return this.row.get();
    }

    @Override
    public double getResolution() {
        return this.row.getResolution();
    }

    @Override
    public boolean isAuthoritative() {
        return this.row.isAuthoritative();
    }

    @Override
    public double getCE() {
        return this.row.getCE();
    }

    @Override
    public double getLE() {
        return this.row.getLE();
    }

    @Override
    public String getUri() {
        return this.row.getUri();
    }

    @Override
    public String getType() {
        return this.row.getType();
    }

    @Override
    public Geometry getBounds() {
        return this.row.getBounds();
    }

    @Override
    public int getFlags() {
        return this.row.getFlags();
    }

    @Override
    public boolean moveToNext() {
        if(this.row != null) {
            if(!this.row.moveToNext())
                this.cursors.remove(0);
            Collections.sort(this.cursors, this.order);
        }
        if(this.cursors.isEmpty())
            return false;
        this.row = this.cursors.get(0);
        return true;
    }

    @Override
    public void close() {
        if(this.cursors.isEmpty())
            return;
        for(ElevationSource.Cursor c : this.cursors)
            c.close();
        this.cursors.clear();
        this.row = null;
    }

    @Override
    public boolean isClosed() {
        return this.cursors.isEmpty();
    }
}
